op-node/rollup/driver/state.go

```diff
@@ -278,11 +280,16 @@ func (s *Driver) eventLoop() {
 				return
 			}
 			if s.network != nil && payload != nil {
-				// Publishing of unsafe data via p2p is optional.
-				// Errors are not severe enough to change/halt sequencing but should be logged and metered.
-				if err := s.network.PublishL2Payload(s.driverCtx, payload); err != nil {
-					s.log.Warn("failed to publish newly created block", "id", payload.ID(), "err", err)
-					s.metrics.RecordPublishingError()
+				// CommitUnsafePayload will fail if this node is not the leader sequencer.
+				if err := s.sequencerConductor.CommitUnsafePayload(s.driverCtx, payload); err != nil {
+					s.log.Error("failed to commit unsafe payload to conductor log", "id", payload.ID(), "err", err)
```

Comment:
I can't access the google doc linked, but if the sequencer is always running and building unsafe payloads that it imports into geth but just not publishing them unless the conductor allows it, won't that lead to the different sequencer nodes all building their on separate unsafe forks (they'd ignore gossip because it doesn't build on their unsafe head)?
Strictly speaking, importing the new unsafe block to geth would be ok (though we want to move that to only happen after gossiping - https://github.com/ethereum-optimism/optimism/pull/8905), the key would be to prevent setting it as the new head via the `engine_forkChoiceUpdated` call unless the conductor allowed it.

Reply:
You're absolutely right here, the goal is like what you described, we want to only import unsafe block to geth after committing to conductor (gossip as well).

Initially the plan was to figure out a way to prevent that, but seems like #8905 already covered it which is great news

Reply:
I don't think that #8905 will actually help you, it changes the order of gossiping but you're just skipping gossiping if the conductor says no and then carrying on with everything else.  The two PRs are changing the bit of logic and are doomed to step on each other's toes though.

Reply:
The end state would be that if we cannot commit to conductor, then do not proceed with all the steps after (gossiping, update fork choice, etc), instead, retry it until commit succeeds or it is no longer sequencing (stopped by conductor).

Seems like change is needed to achieve that

Reply:
Also to answer your previous question

> but if the sequencer is always running and building unsafe payloads that it imports into geth but just not publishing them unless the conductor allows it, won't that lead to the different sequencer nodes all building their on separate unsafe forks (they'd ignore gossip because it doesn't build on their unsafe head)?

1. only active sequencer tries to commit the payload to conductor (put into raft FSM), and raft will help distribute that to other stand by sequencers
2. Under the scenario that our op-conductor is down, and one of the standby sequencer is elected leader, there will be two active sequencers on the network (current one and the newly elected one). The solution here is to not allow gossip and geth block building if commit to consensus failed, this way even if the current sequencer is active, it's states are known only to itself, and does not affect the network

Reply:
Thanks for raising this, Adrian. 👋 howdy, I'm authoring #8905 so I wanted to put some detail on the Gossip Before Import feature here, as I suspect it really messes up your plans.

Gossip Before Import uses a new component to start the payload gossiping to the network *well before* this conductor call would execute. Slightly above this section in this file (line 277, `RunNextSequencerAction`), when *getting* the payload from the engine, also sends down an Async Gossiper to begin gossiping the payload as soon as it is created and sanity tested. This happens way down in the internal `confirmPayload` function, before the block is even reinserted to the engine.

This was done in order to begin gossiping optimistically before the engine does more blocking work (because that work is irrelevant to gossiping). The payload is saved inside the Async Gossiper so that if the block import fails, it can be tried again without rebuilding (and therefore gossiping) a conflicting block.

If these two features were both merged, we'd find that even when the conductor prevents gossiping, the payloads would have already made it out, as the async gossiper begins its work before Line 277 returns.

(Side Note: my PR does not remove the payload publishing in the state loop because I originally thought it would be used somehow by other code paths. I no longer know *why* I thought this, and so my PR will soon feature total removal of this publishing block)

To resolve the conflict, here are the constraints as I understand them:
* Gossip Before Import *must* happen between when the payload is made, and when it is reinserted, which currently happens in the internal `confirmPayload` function.
* Conductor Commit *must* happen prior to any gossiping, because it needs to prevent inappropriate gossip.

and therefore, here are some options:
* Conductor Commit could be passed down into the `confirmPayload` function to gate the pregossiper
* Conductor Commit could be attached to the `OpNode` and used directly in PublishL2Payload, so that *any* caller, regardless of location. I like this outcome but I don't know if it's feasible
* We could decide that this conductor work is simply more important than Gossip Before Import, and drop that feature in order to remove this challenge.

Let me know what you think, I'd be happy to consider other options. I just struggle to think of how we can Gossip Before Import without starting to gossip before we return to the state loop.

Reply:
Now that I've removed the PublishL2Payload from the state loop in my PR, the path forward looks more clear. The Conductor Commit gate will need to be reloacted to either `confirmPayload` to prevent the gossiper from starting, *or* could be attached directly to the Async Gossiper, and it can block itself.

*either way* I think this is something that I'll need to do on my side once this PR is merged. I agree with Adrian that my feature won't help you much, but I can take care of solving this conflict at least.

Reply:
Regarding constraints, I'd like to add one more which is to `if commit in conductor fails, prevent engine to build on top of the payload and eventually update op-geth about it`

With that said, it seems like option 1 would be the choice? As if gossip failed, we could return error in `confirmPayload` and abort remaining steps? Is there any concerns about this logic?

Reply:
Yeah, I think by making the `sequencerConductor` part of the `AsyncGossiper`, we can use your conditional publishing logic without issue. My current `Gossip` function doesn't return an error, but it could.

With regards to your additional constraint, I am not sure how that would work. Any errors that bubble up to `RunNextSequencerAction` are treated as fatal in the current state loop. We *could* just have the work be dropped, returning early without an error. But I am not sure about what the downstream side effects of this would be. @trianglesphere would be more equipped to evaluate that.

Reply:
Gave this constraint, `if commit in conductor fails, prevent engine to build on top of the payload and eventually update op-geth about it` some more though, specifically as you phrased it above,

> The end state would be that if we cannot commit to conductor, then do not proceed with all the steps after (gossiping, update fork choice, etc), instead, retry it until commit succeeds or it is no longer sequencing (stopped by conductor).

As written today, if you cannot commit to the conductor, *you have already done an update fork choice*. As that happens inside the `RunNextSequencerAction` function, blocking gossip here won't stop other aspects of the payload having already been used.

I guess in this case, the conductor logic does indeed need to make the same journey that the AsyncGossiper does

Reply:
Yes, I went through the code within `confirmPayload`, and it seems like the perfect place to inject conductor logic, if we cannot commit to conductor, we bubble up a `derive.ErrTemporary` or even create a new type of error and allow the sequencer step retry safely

Reply:
My understanding is that this code will only be called with non-nil async gossiper and conductor when it is the leader. (If it's verifier, then it'll be called with no-op gossiper and conductor).

And if you're the sequencer, then `NewPayload` should never return invalid block error?